### PR TITLE
graph mode: more docs for insert observers pass

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39828 [wip] qat graph mode: add combine_conv_bn pass
* #39837 eager qat: refactor conv-bn to make conv with scaled weights a func
* #39840 qat graph mode: make per channel observers scriptable
* #39796 qat graph mode: initial scaffolding for prepare_qat_script
* #39795 graph mode: add quantization type enum
* #39750 graph mode qat: make fake_quantize scriptable
* **#39739 graph mode: more docs for insert observers pass**
* #39683 graph mode: add docs to pre-calibration passes

Summary:

Adding more docs and examples to make code reasing easier for newcomers.

Test Plan:

CI, no logic changes

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21975878](https://our.internmc.facebook.com/intern/diff/D21975878)

### CODE REVIEW

File: torch/csrc/jit/passes/quantization/insert_observers.cpp [hunk 1]
```diff
@@ -341,8 +345,12 @@ class InsertObserversHelper {
       const Module& observer_module,
       NameModuleVector& observer_name_and_modules);
 
+  // Uses the state created by fillBoundaryValueMap and fillValueObserverMap
+  // to return an observer for a value, if it is needed.
   c10::optional<Module> getObserverFor(Value* v);
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_observers.cpp:350]: you can change this to `get observer module configured for the value`

File: torch/csrc/jit/passes/quantization/insert_observers.cpp [hunk 2]
```diff
@@ -305,6 +305,10 @@ class InsertObserversHelper {
       bool is_entry_point = false,
       bool is_user_defined_function = false);
 
+  // Record v as "ready for observation" by storing it in values_to_observe and
+  // block_observed_values. If v is a part of a delayed observation pattern,
+  // record v's descendant (per delay rules) instead. The observers are inserted
+  // at a later stage by reading the state created by this function.
   void recordObserved(
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_observers.cpp:312]: value_to_observe stores the values that we'll need to insert observers for, block_observed_values are the values already observed in the current block

File: torch/csrc/jit/passes/quantization/insert_observers.cpp [hunk 3]
```diff
@@ -374,28 +382,75 @@ class InsertObserversHelper {
...
   std::unordered_set<Graph*> visited_graph_of_observer_map_;
+
+  // example:
+  //   %x1 = conv(%x0)
+  //
+  // observer_for_value_ = {
+  //   %x0: FooObserverModule,
+  //   %x1: BarObserverModule,
+  // }
   std::unordered_map<Value*, Module> observer_for_value_;
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_observers.cpp:410]: not sure if example is helpful for this one, I think you can say this is the map from a value to the observer module configured by user.

File: torch/csrc/jit/passes/quantization/insert_observers.cpp [hunk 4]
```diff
@@ -374,28 +382,75 @@ class InsertObserversHelper {
...
   // key is the output of the op, value is a vector of values that need
   // to be observed in order to pass the observed property to the output
+  //
+  // example:
+  //   %x1 = flatten(%x0) // pass_through
+  //   %x2 = conv(%x1) // not pass_through
+  //
+  // pass_through_value_map_ = {
+  //   %x1: [%x0],
+  //   %x2: [],
```
Comment (jerryzh168) [inline torch/csrc/jit/passes/quantization/insert_observers.cpp:450]: we don't have an entry for `%x2` I think

**Review by jerryzh168 — APPROVED** [review]
see some comments inline

File: torch/csrc/jit/passes/quantization/insert_observers.cpp (hunk 3 above)
Comment (vkuzo) [reply torch/csrc/jit/passes/quantization/insert_observers.cpp:410]: agreed, it's probably not that helpful to people who work on this code full time.  It's targeted more to newcomers / people who need to sometimes touch this code.

Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_observers.cpp:410]: I mean the other examples look good, but this one is probably not as helpful as the others

**Pull request closed without merging**